File: Misc/Csamples/dialog_nonmodal1.c

/*

    A non-modal dialog in GTK+. There is also an example of this in the Python folder that is named
similarly. 

    gcc -Wall dialog_nonmodal1.c -o dialog_nonmodal1 `pkg-config --cflags --libs gtk+-3.0`

    Tested on Ubuntu14.04 and GTK3.10

    C. Eric Cashon

*/

#include <gtk/gtk.h>

static void open_dialog(GtkWidget *widget, gpointer data);

int main(int argc, char **argv)
  {      
    gtk_init(&argc, &argv);   

    GtkWidget *window=gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_position(GTK_WINDOW(window), GTK_WIN_POS_CENTER);
    gtk_window_set_default_size(GTK_WINDOW(window), 300, 100);
    gtk_window_set_title(GTK_WINDOW(window), "Non Modal Dialogs");
    g_signal_connect(window, "destroy", G_CALLBACK(gtk_main_quit), NULL);

    GtkWidget *button1=gtk_toggle_button_new_with_label("Show Dialog");
    gtk_widget_set_hexpand(button1, TRUE);
    gtk_widget_set_vexpand(button1, TRUE);
    g_signal_connect(button1, "clicked", G_CALLBACK(open_dialog), window);
      
    GtkWidget *grid=gtk_grid_new();
    gtk_grid_attach(GTK_GRID(grid), button1, 0, 0, 1, 1);

    gtk_container_add(GTK_CONTAINER(window), grid);

    gtk_widget_show_all(window);                  
    gtk_main();
    return 0;
  }
static void open_dialog(GtkWidget *widget, gpointer data)
  {
    //Non modal dialog. Open several dialog windows from the main window.
    static gint i=1;
    gchar *dialog_label=g_strdup_printf("Dialog Window%i", i);
    GtkWidget *dialog=gtk_dialog_new_with_buttons("Dialog", GTK_WINDOW(data), GTK_DIALOG_DESTROY_WITH_PARENT, "Close", GTK_RESPONSE_CLOSE, NULL);
    gtk_window_set_default_size(GTK_WINDOW(dialog), 200, 100);
                                 
    GtkWidget *box=gtk_dialog_get_content_area(GTK_DIALOG(dialog));
    GtkWidget *label=gtk_label_new(dialog_label);
    gtk_box_pack_start(GTK_BOX(box), label, TRUE, TRUE, 0);
    
    g_signal_connect_swapped(dialog, "response", G_CALLBACK(gtk_widget_destroy), dialog);
    gtk_widget_show_all(dialog);

    //Move the windows.
    gint width = gdk_screen_width();
    gint height = gdk_screen_height();
    gint sections_x = width / 100;
    gint sections_y = height / 100;
    gint s1 = (i - 1) % (sections_x * sections_y);
    gint move_x = (s1 % sections_x) * 100;
    gint move_y = (s1 / sections_x) * 100;    
    g_print("Move %i %i %i %i %i %i\n", width, height, sections_x, sections_y, move_x, move_y);
    gtk_window_move(GTK_WINDOW(dialog), move_x, move_y);

    i++;
    g_free(dialog_label);        
  }

File: Misc/Python/dialog_nonmodal1.py
#!/usr/bin/python3

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, Gdk

'''
    A non-modal dialog in GTK+. There is also an example of this in the Csamples folder that is named
similarly.

    C. Eric Cashon

'''

class MainWindow(Gtk.Window):
    def __init__(self):
        Gtk.Window.__init__(self)
        self.set_title("Non Modal Dialogs")
        self.set_default_size(300, 100);
        self.set_position(Gtk.WindowPosition.CENTER)

        self.i = 1 

        button = Gtk.Button("Show Dialog")
        button.connect('clicked', self.show_dialog) 

        self.add(button)

    def show_dialog(self, widget):
        dialog = DialogWindow(self)

class DialogWindow(Gtk.Dialog):
    def __init__(self, main_window):
        Gtk.Dialog.__init__(self, "Dialog", main_window, Gtk.DialogFlags.DESTROY_WITH_PARENT, ("Cancel", Gtk.ResponseType.CANCEL))
        self.set_default_size(200, 100);

        string = "Dialog " + str(main_window.i)
        label = Gtk.Label(string)

        box = self.get_content_area()
        box.pack_start(label, True, True, 0)

        self.connect("response", lambda *x: self.destroy())

        #Move the dialog windows.
        width = Gdk.Screen.width();
        height = Gdk.Screen.height();
        sections_x = width // 100;
        sections_y = height // 100;
        s1 = (main_window.i - 1) % (sections_x * sections_y);
        move_x = (s1 % sections_x) * 100;
        move_y = (s1 // sections_x) * 100;    
        print("Move " +  str(width) + " " + str(height) + " " + str(sections_x) + " " + str(sections_y) + " " + str(move_x) + " " + str(move_y));
        self.move(move_x, move_y); 
        main_window.i += 1

        self.show_all()

win = MainWindow()
win.connect("delete-event", Gtk.main_quit)
win.show_all()
Gtk.main()

